Update Policy Provider

akashsinghal opened the pull request:
- Adds factory and updates configuration to dynamically select policy provider from config
- Updates config policy provider to register as a provider
- Adds `OverallVerifySuccess` method to API and adds implementation in config policy provider
- Adds tests for all additions
- Updates config.json and configmap.yaml to reflect new policy provider configuration

Fixes #152

akashsinghal commented:
@susanshi @mnltejaswini @etrexel would appreciate your feedback on this PR. Thanks!

susanshi commented on config/config.json:
```diff
@@ -9,10 +9,11 @@
     },
     "policies": {
         "version": "1.0.0",
-        "artifactVerificationPolicies": {
```
given this is a significant change to the policy structure, should bump the version? What is our backwards compat requirement,  or is this something we only need to consider after stable release?

akashsinghal replied:
good point. I'm not sure. I just left it as is. I can bump it up, but honestly I don't think what we had before was really a previous version since it basically wasn't used.

susanshi commented on pkg/policyprovider/configpolicy/configpolicy_test.go:
```diff
@@ -80,3 +86,168 @@ func TestPolicyEnforcer_ContinueVerifyOnFailure(t *testing.T) {
+				},
+			},
+			output: true,
+		},
```
Can we add a test case for cosign signature? how does the configuration work when customer is sharing a configuration between cosign/notary

susanshi replied:
or maybe this is something we need to call out that we don't support?

akashsinghal replied:
At the policy level, the customer can specify artifact policy for the notary and cosign signature together. This would mean for an overall success, at least one cosign and notary signature would have to be provided. But then the subject image would need to be a public registry since currently if cosign enabled is set to true, it'll return a 401. I definitely agree we need to call this mixed cased and private registry cosign verification in general. I was thinking of adding a warning here: https://github.com/deislabs/ratify/blob/main/plugins/verifier/cosign/README.md

susanshi commented on pkg/policyprovider/configpolicy/configpolicy_test.go:
```diff
@@ -80,3 +86,168 @@ func TestPolicyEnforcer_ContinueVerifyOnFailure(t *testing.T) {
+					"application/vnd.cncf.notary.v2": "all",
+				},
+			},
+			verifierReports: []interface{}{
```
is there a test case for notary required, but no artifact available?

akashsinghal replied:
I missed this case. I'll add it.

susanshi reviewed (commented):
thanks for moving policy definition to a plugin! left some questions and comments.

akashsinghal commented on pkg/policyprovider/configpolicy/configpolicy_test.go:
```diff
@@ -13,30 +13,36 @@ See the License for the specific language governing permissions and
+		"artifactVerificationPolicies": map[string]types.ArtifactTypeVerifyPolicy{
 			"application/vnd.cncf.notary.v2": "any",
 			"org.example.sbom.v0":            "all",
-			"default":                        "any",
```
that was not intentional. I'll add it back

susanshi commented on pkg/policyprovider/configpolicy/configpolicy_test.go:
```diff
@@ -80,3 +86,168 @@ func TestPolicyEnforcer_ContinueVerifyOnFailure(t *testing.T) {
+					ArtifactType: "application/vnd.cncf.notary.v2",
+				},
+			},
+			output: true,
```
there has been some discussion on fail open or fail close. Looks like in these examples, we are failing open. Should this be property that can be configured?

akashsinghal replied:
yeah that's a good point. I guess it will only be applicable in the scenario where no policies are specified. Right now it'll fail open. The only concern is that will this cause configuration confusion if we add another field? Would the user understand what the extra setting is actually specifying? (Maybe we set a default value for it so it isn't specified for most users)

susanshi replied:
Jimmy mentioned the gatekeeper can be configured to specify fail open vs close. We are property is similar enough , customer might be able to pickup the pattern easily. Could you look into how to specify that in opa/gatekeeper?

akashsinghal replied:
sounds good. I'll look into it

susanshi replied:
thanks @akashsinghal, this thread might have some relevant info. https://open-policy-agent.github.io/gatekeeper/website/docs/failing-closed/

akashsinghal replied:
Discussed offline that we can consider adding a `failurePolicy` to the config in the future https://github.com/deislabs/ratify/issues/167

mnltejaswini commented on charts/ratify/templates/configmap.yaml:
```diff
@@ -30,10 +30,13 @@ data:
-          }
+        "version": "1.0.0",
+        "policyPlugin": {
+            "name": "configPolicy",
```
nit: as it already in the context of policy, do you need the suffix policy?

akashsinghal replied:
I'll remove the policy suffix

mnltejaswini commented on docs/policy-provider.md:
```diff
@@ -0,0 +1,32 @@
+- The `name` field is REQUIRED and MUST match the name of the registered policy provider
+- `artifactVerificationPolicies`: map of artifact type to policy; each entry in the map's policy must be satisfied for Ratify to return true.
+    - `any`: policy that REQUIRES at least one artifact of specified type to verify to `true` 
+    - `all`: policy that REQUIRES all artifacts of specified type to verify to `true`
```
Would be good to call out default behaviors if any

mnltejaswini commented on pkg/policyprovider/factory/factory.go:
```diff
@@ -0,0 +1,73 @@
+// Register adds the factory to the built in providers map
+func Register(name string, factory PolicyFactory) {
+	if factory == nil {
+		panic("store factory cannot be nil")
```
nit: policy factory?

mnltejaswini commented on pkg/policyprovider/configpolicy/configpolicy.go:
```diff
@@ -33,29 +35,56 @@ type PolicyEnforcer struct {
 	}
-	if policyEnforcer.ArtifactTypePolicies["default"] == "" {
-		policyEnforcer.ArtifactTypePolicies["default"] = vt.AnyVerifySuccess
+	if policyEnforcer.ArtifactTypePolicies[defaultPolicyName] == "" {
```
Can people provide default policy in the configuration?

mnltejaswini commented on pkg/executor/core/executor.go:
```diff
@@ -121,13 +115,7 @@ func (executor Executor) verifySubjectInternal(ctx context.Context, verifyParame
-			break
-		}
-	}
+	overallVerifySuccess := executor.PolicyEnforcer.OverallVerifyResult(ctx, verifierReports)
```
So we don't short circuit the execution if any of the verification result fails when the policy is set to all right?

akashsinghal replied:
Yes you're right. I missed this. I suggest that we make the `ContinueVerifyOnFailure` a more generic method called `ContinueVerify` which can determine both in the success and failure case of the current verifier report if it should proceed. That way it can short circuit on first failure for 'all' policy and first success for 'any'. Does it make sense?

akashsinghal replied:
We currently don't short circuit on success for the 'any' policy for that particular artifact type. There is redundant checking that occurs for that particular artifact type if there's already been a success. I think to fix this we would need to keep track of which artifact types we've seen and evaluated to true already...

mnltejaswini commented on pkg/policyprovider/configpolicy/configpolicy.go:
```diff
@@ -74,3 +104,43 @@ func (enforcer PolicyEnforcer) ErrorToVerifyResult(ctx context.Context, subjectR
+
+	for _, report := range verifierReports {
+		castedReport := report.(verifier.VerifierResult)
+		// the artifact type of the verified artifact matches type specified in policy
```
If all verifications for a type that is not set in the config (with default policy)  fails, what is the overall result?

akashsinghal replied:
Currently OverallVerifyResult only considers the artifact types that are explicitly specified. And those are the only ones that are taken into account for overall success. So it disregards the success/failure of an unspecified artifact type.

mnltejaswini commented on pkg/policyprovider/configpolicy/configpolicy.go:
```diff
@@ -74,3 +104,43 @@ func (enforcer PolicyEnforcer) ErrorToVerifyResult(ctx context.Context, subjectR
+				verifySuccess[castedReport.ArtifactType] = true
+			} else if policyType == vt.AllVerifySuccess {
+				// if policy is 'all'
+				if !castedReport.IsSuccess {
```
Wodnering if this statement will ever be hit? AS this should be caught in the ContinueVerifyOnFailure right?

akashsinghal replied:
yeah... it won't be hit with the current executor set up. Should I remove it? I wasn't sure if the implementation of this method should be independent of its usage in the executor.

akashsinghal commented:
@susanshi @mnltejaswini I made some changes to the configPolicy Provider:
1. I changed the default policy behavior a bit. The default policy is applied to and enforced for any artifacts that are encountered by the verifier but not specified explicitly in the artifactTypePolicies section of the policy configuration. Previously, only artifacts of the artifact type specified in the config were considered in the OverallVerifySuccess. It didn't matter if other artifacts were verified true/false. Now it will take unspecified artifact types into consideration by giving it a default policy.
2. The default policy introduces a new policy type called 'none'. This requires 0+ artifacts of a type to verify to true to satisfy the policy. This is the default value for the 'default' policy. With this 'none' policy, unspecified artifacts are by default ignored in the verification process. The user can override this 'default' policy and set to, for example, 'any' which will then enforce at least one artifact type success for every unspecified artifact encountered. 
3. The `VerifyNeeded` is now being used. It now checks if an artifact's type policy exists in the policy configuration as 'any' or 'all' and it will ONLY allow verification of artifacts that are of policy type 'any' or 'all'. 
4. The executor had to be changed slightly to accommodate 0 verifier reports scenarios. Previously if there were no verifier reports that resulted, executor assumed we were unable to extract referrers from the subject. Now that's not a valid assumption so the 0 referrers check is done immediately after the referrers are queried from the registry using ORAS. 
5. I added tests for `VerifyNeeded` and updated executor and config policy tests to accommodate the new changes.

susanshi commented on docs/policy-provider.md:
```diff
@@ -0,0 +1,50 @@
+...
+"policies": {
+    "version": "1.0.0",
+    "plugin": {
```
adding a comment to make sure we are updating the config at https://github.com/deislabs/ratify-action/blob/main/config.json

susanshi commented:
> 



> configPolicy



> @susanshi @mnltejaswini I made some changes to the configPolicy Provider:
> 
> 1. I changed the default policy behavior a bit. The default policy is applied to and enforced for any artifacts that are encountered by the verifier but not specified explicitly in the artifactTypePolicies section of the policy configuration. Previously, only artifacts of the artifact type specified in the config were considered in the OverallVerifySuccess. It didn't matter if other artifacts were verified true/false. Now it will take unspecified artifact types into consideration by giving it a default policy.
> 2. The default policy introduces a new policy type called 'none'. This requires 0+ artifacts of a type to verify to true to satisfy the policy. This is the default value for the 'default' policy. With this 'none' policy, unspecified artifacts are by default ignored in the verification process. The user can override this 'default' policy and set to, for example, 'any' which will then enforce at least one artifact type success for every unspecified artifact encountered.
> 3. The `VerifyNeeded` is now being used. It now checks if an artifact's type policy exists in the policy configuration as 'any' or 'all' and it will ONLY allow verification of artifacts that are of policy type 'any' or 'all'.
> 4. The executor had to be changed slightly to accommodate 0 verifier reports scenarios. Previously if there were no verifier reports that resulted, executor assumed we were unable to extract referrers from the subject. Now that's not a valid assumption so the 0 referrers check is done immediately after the referrers are queried from the registry using ORAS.
> 5. I added tests for `VerifyNeeded` and updated executor and config policy tests to accommodate the new changes.

@akashsinghal , maybe these internal working should be captured in docs/policyplugin?

susanshi commented on docs/policy-provider.md:
```diff
@@ -0,0 +1,50 @@
+    - `none`: policy that does NOT require any artifacts to verify to `true`
+- Default policy:
+    - The `default` policy applies to unspecified artifact types. The `default` policy is set to `none`. Thus, all unspecified artifact types will be ignored in determing overall success result.
+    - The `default` policy can be overriden to `any` or `all` in the map:
```
Suggestion to provide some examples for any. If 'any' is configured, does it required 1+ of a specific type to be valid to produce a success result? or does the artifact type not matter

akashsinghal replied:
Added some examples demonstrating. Could you take another look at them @susanshi. Not sure if they are confusing.

susanshi commented on docs/policy-provider.md:
```diff
@@ -0,0 +1,50 @@
+
+Currently, Ratify supports a Configuration based Policy Provider named `configPolicy`.
+
+## Sample Config
```
Should this be documentation for default configuration

akashsinghal replied:
I think we might need to discuss what we want the default config to be now. Since we have default policy support of 'all'. we can technically omit the `artifactTypePolicies` section entirely. So config would be:  
``` ...
    "policies": {
        "version": "1.0.0",
        "plugin": {
            "name": "configPolicy"
        }
    },
    ...
```

susanshi replied:
ah I see. In the examples, we can still specify them for readability, but call out they can be omitted as they are the default. Can we also add a note about if there are artifacts that cannot be verified, how does it impact the overall result?

akashsinghal commented:
@mnltejaswini @susanshi I've updated the policy provider:
- I removed the `none` policy like we discussed in the meeting. I also made the default policy `all`
- Now all unspecified artifact types will have default policy 'all' which means all artifacts NEED to be verified to true for overall verify success to be true.
- I updated the tests accordingly to match this new behavior
- I also updated the documentation to include some examples.

sajayantony commented:
I think default all might be good to start and merge this. Going forward if we iron out the requirements for fail open we could open issue to track that.

akashsinghal commented:
We should address adding ability for users to specify the artifact type policies in the values.yaml file cc: @etrexel

akashsinghal commented on pkg/policyprovider/configpolicy/configpolicy_test.go:
```diff
@@ -80,3 +87,221 @@ func TestPolicyEnforcer_ContinueVerifyOnFailure(t *testing.T) {
+				"artifactVerificationPolicies": map[string]types.ArtifactTypeVerifyPolicy{},
+			},
+			verifierReports: []interface{}{},
+			output:          true,
```
we could change this behavior to return true in the OverallVerifySuccess method. however, in the executor, if there are no referrers found, it'll actually throw an error before it reaches the the OverallVerifySuccess method invocation. https://github.com/akashsinghal/ratify/blob/be813bd32219038340cb0d8c919fc4194fdfcf18/pkg/executor/core/executor.go#L115

akashsinghal replied:
since an error is being thrown by the executor anyways. I think it makes sense for this to return false. I'll change it

akashsinghal commented on docs/policy-provider.md:
```diff
@@ -0,0 +1,90 @@
+    - `any`: policy that REQUIRES at least one artifact of specified type to verify to `true` 
+    - `all`: policy that REQUIRES all artifacts of specified type to verify to `true``
+- Default policy:
+    - The `default` policy applies to unspecified artifact types. The `default` policy is set to `all`. Thus, all unspecified artifact types must have all successful verification results for an overall success result.
```
This one's a bit more nuanced unfortunately. The `any` and `all` policies only apply after we go through the verification process and have our verifier reports. The OverallVerifyResult method gets called last with the reports passed. If we cannot find a verifier that can verify a specific artifact type, it's essentially ignored right now. There is a check here: https://github.com/akashsinghal/ratify/blob/be813bd32219038340cb0d8c919fc4194fdfcf18/pkg/executor/core/executor.go#L128 which only continues verification for the artifact if there is a verifier it finds that can verify that `artifactType`.

akashsinghal replied:
Added a section on where/how the policy provider is invoked during execution in executor

susanshi reviewed (approved):
thanks for the clarification @akashsinghal. A separate item tracks [override policy in template](https://github.com/deislabs/ratify/issues/177). Feel free to merge  after doc updates.

akashsinghal commented:
@susanshi I updated the policy provider doc to discuss the verifier behaviors we were discussing. I also added an executor [doc](https://github.com/deislabs/ratify/blob/f5a822359fb7ed5d9a2d70333ecbe376456cedb2/docs/executor.md)  walking through all logic implemented so far.
